### PR TITLE
Clean up variable API usage in HMC and NUTS

### PR DESCRIPTION
**Pull request by neerajprad**
This cleans up the HMC/NUTS implementations:
 - Removes usage of `Variable/variable` in favor of `torch.tensor`.
 - Removes `requires_grad=True` from latent sites. Instead, we set the flag right before taking derivatives inside `integrator._grad`, and unset it right after.

NOTE: This implements @gchanan's suggestion from #869.

### CODE REVIEW

File: pyro/ops/integrator.py [hunk 1]
```diff
@@ -59,9 +59,11 @@ def single_step_velocity_verlet(z, r, potential_fn, step_size, z_grads=None):
 
 
 def _grad(potential_fn, z):
-    z = {k: Variable(v, requires_grad=True) for k, v in z.items()}
     z_keys, z_nodes = zip(*z.items())
+    for node in z_nodes:
```
Comment (neerajprad) [inline pyro/ops/integrator.py]: @apaszke, @gchanan - As mentioned in #869, I am setting `requires_grad=True` here right before calling `.grad` and then unsetting it later. If we set this to `True` at the time of tensor declaration, we can just call `.grad` directly here without having to set/unset these flags, but that ends up recording a lot of other tensor operations that we don't need to and is much slower as a result.

File: pyro/infer/mcmc/hmc.py [hunk 2]
```diff
@@ -189,7 +189,7 @@ def cleanup(self):
         self._reset()
 
     def sample(self, trace):
-        z = {name: node["value"] for name, node in trace.iter_stochastic_nodes()}
+        z = {name: node["value"].data for name, node in trace.iter_stochastic_nodes()}
```
Comment (fritzo) [inline pyro/infer/mcmc/hmc.py]: Should this be `.detach()` rather than `.data`? I'm not sure the difference, and I thought we should always use `.detach()` now.

File: tests/infer/mcmc/test_hmc.py [hunk 3]
```diff
@@ -25,20 +24,20 @@ def __init__(self, dim, chain_len, num_obs):
         self.dim = dim
         self.chain_len = chain_len
         self.num_obs = num_obs
-        self.mu_0 = Variable(torch.zeros_like(torch.Tensor(self.dim)), requires_grad=True)
-        self.lambda_prec = Variable(torch.ones_like(torch.Tensor(self.dim)))
+        self.mu_0 = torch.zeros_like(torch.tensor(self.dim))
```
Comment (fritzo) [inline tests/infer/mcmc/test_hmc.py]: Why not simply `torch.zeros(self.dim)`?

File: tests/infer/mcmc/test_hmc.py [hunk 4]
```diff
@@ -25,20 +24,20 @@ def __init__(self, dim, chain_len, num_obs):
...
+        self.lambda_prec = torch.ones_like(torch.Tensor(self.dim))
```
Comment (fritzo) [inline tests/infer/mcmc/test_hmc.py]: Why not simply `torch.ones(self.dim)`?

File: tests/infer/mcmc/test_hmc.py [hunk 5]
```diff
@@ -140,8 +139,8 @@ def test_hmc_conjugate_gaussian(fixture,
         param_name = 'mu_' + str(i)
         latent_mu = torch.mean(torch.stack(post_trace[param_name]), 0)
         latent_std = torch.std(torch.stack(post_trace[param_name]), 0)
-        expected_mean = Variable(torch.ones_like(torch.Tensor(fixture.dim)) * expected_means[i - 1])
-        expected_std = 1 / torch.sqrt(Variable(torch.ones_like(torch.Tensor(fixture.dim)) * expected_precs[i - 1]))
+        expected_mean = torch.ones_like(torch.tensor(fixture.dim)) * expected_means[i - 1]
```
Comment (fritzo) [inline tests/infer/mcmc/test_hmc.py]: Consider using the cheaper `torch.Tensor(fixture.dim).fill_(expected_means[i - 1])`

File: tests/infer/mcmc/test_hmc.py [hunk 6]
```diff
@@ -25,20 +24,20 @@ def __init__(self, dim, chain_len, num_obs):
...
             mu = pyro.sample('mu_{}'.format(i),
-                             dist.Normal(mu=mu, sigma=Variable(lambda_prec.data)))
-        pyro.sample('obs', dist.Normal(mu, Variable(lambda_prec.data)), obs=data)
+                             dist.Normal(mu=mu, sigma=lambda_prec))
+        pyro.sample('obs', dist.Normal(mu, lambda_prec), obs=data)
 
     @property
     def data(self):
-        return Variable(torch.ones_like(torch.Tensor(self.num_obs, self.dim)))
+        return torch.ones_like(torch.Tensor(self.num_obs, self.dim))
```
Comment (fritzo) [inline tests/infer/mcmc/test_hmc.py]: Why not simply `torch.ones(self.num_obs, self.dim))`?

File: tests/infer/mcmc/test_hmc.py [hunk 7]
```diff
@@ -181,16 +180,16 @@ def model(data):
 
 def test_bernoulli_beta():
     def model(data):
-        alpha = pyro.param('alpha', variable([1.1, 1.1], requires_grad=True))
-        beta = pyro.param('beta', variable([1.1, 1.1], requires_grad=True))
+        alpha = pyro.param('alpha', torch.tensor([1.1, 1.1]))
```
Comment (fritzo) [inline tests/infer/mcmc/test_hmc.py]: Hmm it seems a bit un-idiomatic to call `pyro.param` with `requires_grad=False`. Can we discuss to determine whether `with torch.no_grad()` would be a suitable alternative solution that could (1) preserve our existing speed, but (2) allow more idiomatic use of grad-requiring params?

File: pyro/infer/mcmc/hmc.py (hunk 2 above)
Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Both are actually the same, and return a new instance with requires_grad=True, sharing the same underlying data. Will change to `.detach()` for consistency.

File: tests/infer/mcmc/test_hmc.py (hunk 3 above)
Comment (neerajprad) [reply tests/infer/mcmc/test_hmc.py]: My bad, all of this can be simplified now, since we `torch.zeros/ones` support dytes.

File: tests/infer/mcmc/test_hmc.py (hunk 7 above)
Comment (neerajprad) [reply tests/infer/mcmc/test_hmc.py]: Actually, this was written in the spirit of other SVI models, but there really is no need to use the `pyro.param` primitive here. We could just use regular variables, and that might make things clearer since we are not interacting with the param store at all. Does that sound reasonable?

```
        alpha = torch.tensor([1.1, 1.1])
        beta = torch.tensor([1.1, 1.1])
```

Additionally, we are just generating samples from the latent sites and do not need `requires_grad` to be set to true on these parameter sites at all. So overall, using `pyro.param` just seems to be adding confusion here.

The only down-side (arguable) is that we can't just swap out a model in SVI, and use `HMC` to generate samples from it, but the semantics are sufficiently different anyhow. What do you think?

**Comment by neerajprad** [comment]
> Let's chat about our desired idiom for pyro.param

@fritzo - I removed `pyro.param` from the HMC tests since we don't need to interact with the param store at all for sampling. Lets discuss tomorrow if you have any thoughts or reservations.

File: pyro/ops/integrator.py (hunk 1 above)
Comment (apaszke) [reply pyro/ops/integrator.py]: Yup that looks good

File: tests/infer/mcmc/test_hmc.py (hunk 7 above)
Comment (fritzo) [reply tests/infer/mcmc/test_hmc.py]: Thanks for addressing this. Let's merge and then discuss how to make it easy to swap out SVI for HMC.
